### DISCUSSION TITLE
Event System

### DISCUSSION THREAD

**Discussion by bglamadrid** [trebol-ecommerce/trebol-backend-monolith#164] [Ideas]
## Summary

Introduce a mechanism to create documented events, each fired under certain conditions such as

- A new order is requested by a customer
- Payment of an order is validated against a third-party, effectively rendering it as paid
- A product's current stock reaches zero

And so on. 

These examples should not be considered in particular; the feature I'm conceiving is an abstraction, an API to both fire and listen to any defined event and let other parts of the application (and perhaps outside of the application itself) be aware of these events.

### Bare-bones idea

- Have a singleton Spring service act as a global event emitter (broadcast)
  - It also should provide methods that other services may use to listen to specific events
- Events themselves have to be serialized using the smallest format available to reduce all possible processing time
  - And yet, events could (and probably should) link to context objects of any type

### Expected outcome

Have, at every developer's disposal, a simple yet powerful mean to extend the features of this project.

### Considered alternatives 

I haven't yet searched for alternatives, as I often try to figure out how I would do things myself (no I don't endorse doing this, I personally just can't help it).

**Comment by mepox** [comment]
Well Spring has events so that could be useful in this?
https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#context-functionality-events

**Reply by bglamadrid** [reply]
That certainly seems it could work. I would have to try first though.

**Comment by NyorJa** [comment]
Would you like to consider Kafka in this scenario?

**Reply by bglamadrid** [reply]
I've never used Kafka. Isn't that a tool that operates outside of the scope of a Spring project? If so, perhaps we could provide a Kafka integration when things get big and bad.

**Reply by NyorJa** [reply]
Yeah, its outside. Since trebol-backend is monolithic I think its a good idea to have this on a separate springboot project. Im exploring Kafka also in my personal projects and its a good choice when it comes to event driven system scenario
